<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate summary statistics for an alignment. — calculate_alignment_stats • baitfindR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Calculate summary statistics for an alignment. — calculate_alignment_stats" />

<meta property="og:description" content="Including the original alignment in the output with include_aln
can be useful for mapping calculate_alignment_stats over a list
of alignments with map_df to sort and filter
alignments by their summary statistics." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">baitfindR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/joelnitta/baitfindR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate summary statistics for an alignment.</h1>
    <small class="dont-index">Source: <a href='https://github.com/joelnitta/baitfindR/blob/master/R/calculate_alignment_stats.R'><code>R/calculate_alignment_stats.R</code></a></small>
    <div class="hidden name"><code>calculate_alignment_stats.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Including the original alignment in the output with <code>include_aln</code>
can be useful for mapping <code>calculate_alignment_stats</code> over a list
of alignments with <code><a href='http://www.rdocumentation.org/packages/purrr/topics/map'>map_df</a></code> to sort and filter
alignments by their summary statistics.</p>
    
    </div>

    <pre class="usage"><span class='fu'>calculate_alignment_stats</span>(<span class='no'>alignment</span>, <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fl'>120</span>, <span class='kw'>cutoff_any</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>include_aln</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>alignment</th>
      <td><p>Input alignment; must be a matrix of class "DNAbin".</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p>Numeric value indicating minimum exon length (optional);
flag this alignment if any/all exons are less than the cutoff length.</p></td>
    </tr>
    <tr>
      <th>cutoff_any</th>
      <td><p>Logical; Should the alignment
be flagged if any exons are shorter than the cutoff? The default, FALSE,
means that the alignment will only be flagged if all exons are shorter
than the cutoff value.</p></td>
    </tr>
    <tr>
      <th>include_aln</th>
      <td><p>Logical; Should the original alignment
be included in the output list?</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list including the following summary statistics:</p><dl class='dl-horizontal'>
  <dt>intron_lengths</dt><dd><p>List including vector of intron lengths</p></dd>
  <dt>exon_lengths</dt><dd><p>List including vector of exon lengths</p></dd>
  <dt>num_introns</dt><dd><p>Number of introns</p></dd>
  <dt>num_exons</dt><dd><p>Number of introns</p></dd>
  <dt>mean_dist</dt><dd><p>Mean genetic distance between sequences in alignment</p></dd>
  <dt>max_dist</dt><dd><p>Maximum genetic distance between sequences in alignment</p></dd>
  <dt>GC_content</dt><dd><p>Total %GC content</p></dd>
  <dt>pars_inf</dt><dd><p>Fraction of sites that are parsimony informative</p></dd>
  <dt>total_exon_length</dt><dd><p>Total exon length</p></dd>
  <dt>less_than_cutoff</dt><dd><p>Logical flag indicating whether alignment passed
  the minimum exon length cutoff or not</p></dd>
  <dt>alignment</dt><dd><p>The original input alignment</p></dd>
</dl>

    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Joel H. Nitta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

