<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Infer trees from fasta files. — fasta_to_tree • baitfindR</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Infer trees from fasta files. — fasta_to_tree" />

<meta property="og:description" content="Given a folder containing unaligned sequences in fasta format (i.e., clusters),
aligns each cluster with mafft (small clusters) or pasta (large clusters),
excludes poorly aligned sites with phyutility, and infers a maximum-likelihood
tree with RAxML (small clusters) or fasttree (large clusters). Requires all
of these programs to be installed and included in the user's $PATH. Assumes clusters are named &quot;cluster1.fa&quot;, &quot;cluster2.fa&quot;, etc. Clusters with fewer than 1,000 sequences are considered &quot;small,&quot; and those with more are considered &quot;large.&quot;" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">baitfindR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/translation.html">Translation</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/joelnitta/baitfindR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Infer trees from fasta files.</h1>
    <small class="dont-index">Source: <a href='https://github.com/joelnitta/baitfindR/blob/master/R/yang_and_smith.R'><code>R/yang_and_smith.R</code></a></small>
    <div class="hidden name"><code>fasta_to_tree.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Given a folder containing unaligned sequences in fasta format (i.e., clusters),
aligns each cluster with <code>mafft</code> (small clusters) or <code>pasta</code> (large clusters),
excludes poorly aligned sites with <code>phyutility</code>, and infers a maximum-likelihood
tree with <code>RAxML</code> (small clusters) or <code>fasttree</code> (large clusters). Requires all
of these programs to be installed and included in the user's <code>$PATH</code>. Assumes clusters are named <code>"cluster1.fa"</code>, <code>"cluster2.fa"</code>, etc. Clusters with fewer than 1,000 sequences are considered "small," and those with more are considered "large."</p>
    
    </div>

    <pre class="usage"><span class='fu'>fasta_to_tree</span>(<span class='kw'>path_to_ys</span> <span class='kw'>=</span> <span class='kw pkg'>pkgconfig</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/pkgconfig/topics/get_config'>get_config</a></span>(<span class='st'>"baitfindR::path_to_ys"</span>),
  <span class='no'>seq_folder</span>, <span class='no'>number_cores</span>, <span class='no'>seq_type</span>, <span class='kw'>bootstrap</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>overwrite</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>get_hash</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>echo</span> <span class='kw'>=</span> <span class='kw pkg'>pkgconfig</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/pkgconfig/topics/get_config'>get_config</a></span>(<span class='st'>"baitfindR::echo"</span>, <span class='kw'>fallback</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>), <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>path_to_ys</th>
      <td><p>Character vector of length one; the path to the folder containing Y&amp;S python scripts, e.g., <code>"/Users/me/apps/phylogenomic_dataset_construction/"</code></p></td>
    </tr>
    <tr>
      <th>seq_folder</th>
      <td><p>Character vector of length one; the path to the folder containing the fasta files.</p></td>
    </tr>
    <tr>
      <th>number_cores</th>
      <td><p>Numeric; number of threads to use for <code>RAxML</code> and <code>mafft</code>.</p></td>
    </tr>
    <tr>
      <th>seq_type</th>
      <td><p>Character vector of length one indicating type of sequences. Should either be <code>"dna"</code> for DNA or <code>"aa"</code> for proteins.</p></td>
    </tr>
    <tr>
      <th>bootstrap</th>
      <td><p>Logical; should run a bootstrap analysis be run for the trees?</p></td>
    </tr>
    <tr>
      <th>overwrite</th>
      <td><p>Logical; should previous output of this command be erased so new output can be written? Once erased it cannot be restored, so use with caution!</p></td>
    </tr>
    <tr>
      <th>get_hash</th>
      <td><p>Logical; should the 32-byte MD5 hash be computed for all output tree files concatenated together? Used for by <code><a href='https://www.rdocumentation.org/packages/drake/topics/drake_plan'>drake_plan</a></code> for tracking during workflows. If <code>TRUE</code>, this function will return the hash.</p></td>
    </tr>
    <tr>
      <th>echo</th>
      <td><p>Logical; should the standard output and error be printed to the screen?</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Other arguments. Not used by this function, but meant to be used by <code><a href='https://www.rdocumentation.org/packages/drake/topics/drake_plan'>drake_plan</a></code> for tracking during workflows.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>For each input cluster <code>cluster1.fa</code> in <code>seq_folder</code>, <code>cluster1.fa.mafft.aln</code> (small clusters) or <code>cluster1.pasta.aln</code> (large clusters), <code>cluster1.fa.mafft.aln-cln</code> (small clusters) or <code>cluster1.fa.pasta.aln-cln</code> (large clusters), and <code>cluster1.raxml.tre</code> (small clusters) or <code>cluster1.fasttree.tre</code> (large clusters) will be written to <code>seq_folder</code>. If <code>get_hash</code> is <code>TRUE</code>, the 32-byte MD5 hash be computed for all <code>.tre</code> files concatenated together will be returned.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Wrapper for Yang and Smith (2014) <code>fasta_to_tree.py</code></p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Yang, Y. and S.A. Smith. 2014. Orthology inference in non-model organisms using transcriptomes and low-coverage genomes: improving accuracy and matrix occupancy for phylogenomics. Molecular Biology and Evolution 31:3081-3092. <a href='https://bitbucket.org/yangya/phylogenomic_dataset_construction/overview'>https://bitbucket.org/yangya/phylogenomic_dataset_construction/overview</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'>fasta_to_tree</span>(<span class='kw'>seq_folder</span> <span class='kw'>=</span> <span class='st'>"some/folder/containing/fasta/seqs"</span>, <span class='kw'>number_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"dna"</span>, <span class='kw'>bootstrap</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Joel H Nitta, <a href='mailto:joelnitta@gmail.com'>joelnitta@gmail.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Joel H. Nitta.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

