# Build blast db
build_blast_db <- function (in_seqs, out_name, db_type = "nucl", other_args = "-parse_seqids", outfile) {
  arguments <- paste0("-in ", in_seqs, " -out ", out_name, " -dbtype ", db_type, " ", other_args)
  system2("makeblastdb", arguments)
}

#' blastp_long_orfs
#'
#' Wrapper to call blastp
#'
#' Extracts long open reading frames (ORFs) from a fasta file containing transcript sequences (i.e., the transcriptome).
#'
# Wrapper to call blastp for long orfs generated by transdecoder
# By default other_args produces outformat 6
blastp_long_orfs <- function (query, blast_result_file, database, other_args = "-max_target_seqs 1 -outfmt 6 -evalue 10 -num_threads 2", ...) {
  arguments <- paste0("-query ", query, " -db ", database, " ", other_args, " > ", blast_result_file)
  system2("blastp", arguments)
}

# blast n single file, save as outformat 6
blast_cdhitest <- function (query, blast_result_file, database, other_args = "-evalue 10 -num_threads 1 -max_target_seqs 1000 -outfmt '6 qseqid qlen sseqid slen frames pident nident length mismatch gapopen qstart qend sstart send evalue bitscore'", ...) {
  arguments <- paste0("-query ", query, " -db ", database, " ", other_args, " > ", blast_result_file)
  system2("blastn", arguments)
}
